src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c

```diff
@@ -3539,10 +3546,17 @@ scheduler_service_cores(struct xocl_scheduler *xs)
 	struct list_head *pos, *next;
 
 	SCHED_DEBUGF("-> %s\n", __func__);
+
+	mutex_lock(&xs->scheduler_lock);
 	list_for_each_safe(pos, next, &xs->cores) {
 		struct exec_core *exec = list_entry(pos, struct exec_core, core_list);
+
+		mutex_unlock(&xs->scheduler_lock);
```

Comment:
how can you make sure the exec is not going to be destroyed after you drop the lock here?

Reply:
We cannot, as of today, we don't prevent this situation too. Given the exec will be thrown into the queues, running queue, pending_queue, etc. Seems we don't have a race otherwise we should see exec being destroyed very often?

I think the reset procedure has covered most of the cases to prevent this situation. I am not sure if just add lock around line 3555 would be safe enough.

Reply:
I thought that this is the exact race you're trying to fix, no? If you drop lock here, all your effort goes in vain, right? Can we keep the lock here?

Reply:
After an offline chat, we agreed that keep lock to protect this whole operation. Two additional tests to be done:
1) wait for pipeline to finish
2) I have a specific test procedure which can trigger the scheduler_loop panic sooner(within 10~20 reset cycles), will verify the panic is reproducible without the fix, and cannot be seen with the fix.